meson.build

```diff
@@ -0,0 +1,117 @@
+project('cubic-server', 'cpp',
+  version : '0.1',
+  default_options : ['warning_level=3',
+                     'cpp_std=c++20'])
+
+#======================
+# variables declaration
+#======================
+
+cxxflags =          []
+ldflags  =          []
+cubic_include =     []
+cubic_src =         []
+cubic_deps =        []
+
+#=================================
+# use libc++ on clang if installed
+#=================================
+
+if get_option('libcpp_clang')
+    compiler = meson.get_compiler('cpp')
+    if compiler.get_id() == 'clang'
+        cxxflags += ['-stdlib=libc++']
+        ldflags +=  ['-stdlib=libc++']
+    else
+        warning('Clang only option ignored because used on compiler ' + compiler.get_id())
+    endif
+endif
+
+#=====================
+# set name and version
+#=====================
+
+cxxflags += ['-DPROGRAM_NAME="CubicServer"', '-DPROGRAM_VERSION="' + meson.project_version() + '"' ]
+
+#=====================
+# Manage dependencies.
+#=====================
+
+# pkgconfig is required for meson to work correctly.
+# dependencies might not be found without it on linux.
+
+thread = dependency('threads',   required: true)
+gtkmm =  dependency('gtkmm-3.0', required: true)
+curl =   dependency('libcurl', required: true)
+
+yaml     = dependency('yaml-cpp',      required: false)
+json     = dependency('nlohmann_json', required: false)
+noise    = dependency('perlin_noise',  required: false)
+argparse = dependency('argparse',      required: false)
+crcpp    = dependency('CRCpp',         required: false)
+
+if get_option('buildtype') == 'debug' and get_option('test')
+  gtest = dependency('gtest', main: true, required: false)
+  rapidcheck = dependency('rapidcheck', required: false)
+endif
+
+#===============================
+# retrieve missing dependencies.
+#===============================
+
+cmake = import('cmake')
+
+if not yaml.found()
+  yaml_subproj = cmake.subproject('yaml-cpp')
+  yaml = yaml_subproj.dependency('yaml-cpp')
+endif
+
+if not json.found()
+  json = dependency('nlohmann_json')
+endif
```

Comment:
Why do you first check for it with required: false, then check for it again with required: true if it is not found the first time? Just remove "required: false" from the first lookup. :)

Reply:
At first, it was so I could choose between using system deps and wrap system. I will do this for the heaviest dependencies (boost i'm looking at you), with the default being recompiling boost if needed, otherwise require a system installation. I don't know if that could be considered good practice?

Reply:
How does this let you choose between the two? Note that the default is already to try finding a system dependency first, and if permitted, fall back to a wrap dependency (by default if the dependency is required it can fallback, and cannot fallback if it's optional; this can be controlled with the `allow_fallback` kwarg).

If users want to enforce either choice, it's possible to do so by configuring the build with the option `--wrap-mode=` and the possible values:
- nofallback -- never use wraps
- forcefallback -- never use system dependencies 
- default
- ...

You can also force fallback for exactly a specific set of dependencies with e.g. `--force-fallback-for=boost, perlin_noise`.

Reply:
You can at least limit the `required: false` to the cases where you need to specially handle it with cmake. :)

Reply:
indeed! that's what i'll do at first :) I will still try to create patches on the side as I feel it's good practice

Reply:
Also yes, I'm very happy we (meson) were able to design a system that puts users and builders in control of how vendored vs. system dependencies works... It's something that has consistently bothered me in every single build system other than meson, and I've used a lot of them (not just autotools and cmake).

Reply:
true, I have only two things that bothers me with it: not enough dependency are using meson (so more cmake wrapper haha)
and the fact you need to solve each dependency one at a time if they are missing, it does not tell "hey, I need those 4 dependencies but you don't have 3 of them, come back".
Since I am trying to port this project from cmake to meson, I clearly see the difference in readability, and in pushed good practice. I hope it will not become a cluster of legacy things as cmake did :/
autotools is just (sadly) the worst I tried, too slow, checking for useless things...
scons... well scons.
I really like this way of managing dependencies tho, just saying "find it, if you don't, it's there" making the use of another program for dependency management useless is perfect (no more conan, nuget...)